pkg/webhook/resources/node/validator.go

```diff
@@ -71,3 +97,139 @@ func validateCordonAndMaintenanceMode(oldNode, newNode *corev1.Node, nodeList []
 	}
 	return werror.NewBadRequest("can't enable maintenance mode or cordon on the last available node")
 }
+
+func (v *nodeValidator) validateCPUManagerOperation(node *corev1.Node) error {
+	annot, ok := node.Annotations[util.AnnotationCPUManagerUpdateStatus]
+	if !ok {
+		return nil
+	}
+
+	updateStatus, err := ctlnode.GetCPUManagerUpdateStatus(annot)
+	if err != nil {
+		return werror.NewBadRequest(fmt.Sprintf("Failed to retrieve cpu-manager-update-status from annotation: %v", err))
+	}
+	// only validate when update status is requested
+	if updateStatus.Status != ctlnode.CPUManagerRequestedStatus {
+		return nil
+	}
+	policy := updateStatus.Policy
+
+	// check if cpu manager policy is the same
+	if err := checkCPUManagerLabel(node, policy); err != nil {
+		return err
+	}
+	// check if there is other job that still updating cpu manager policy to the same node
+	if err := checkCurrentNodeCPUManagerJobs(node, v.jobCache); err != nil {
+		return err
+	}
+	// check if this node is master and there are other master nodes undating cpu manager policy
+	// since the policy update on master node need to restart rke2-server, we only allow one master node update policy
+	if err := checkMasterNodeJobs(node, v.nodeCache, v.jobCache); err != nil {
+		return err
+	}
+	// check if there is any vm that enable cpu pinning while cpu manager is going to be disabled
+	if err := checkCPUPinningVMIs(node, policy, v.vmiCache); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkCPUManagerLabel(node *corev1.Node, policy ctlnode.CPUManagerPolicy) error {
+	cpuManagerLabel, err := strconv.ParseBool(node.Labels[kubevirtv1.CPUManager])
+	// means CPUManager feature gate not enabled
+	if err != nil {
+		return werror.NewBadRequest("label cpumanager not found")
+	}
+	// check if the cpu manager policy is the same
+	if ((policy == ctlnode.CPUManagerStaticPolicy) && cpuManagerLabel) || ((policy == ctlnode.CPUManagerNonePolicy) && !cpuManagerLabel) {
+		return werror.NewBadRequest(fmt.Sprintf("current cpu manager policy is already the same as requested value: %s", policy))
+	}
+	return nil
+}
+
+func checkCurrentNodeCPUManagerJobs(node *corev1.Node, jobCache ctlbatchv1.JobCache) error {
+	jobNames, err := getCPUManagerRunningJobNamesOnNodes(jobCache, []string{node.Name})
+	if err != nil {
+		return werror.NewInternalError(err.Error())
+	}
+	if len(jobNames) > 0 {
+		return werror.NewBadRequest(fmt.Sprintf("there is other job %s updating the cpu manager policy for this node %s", strings.Join(jobNames, ", "), node.Name))
+	}
+	return nil
+}
+
+func checkMasterNodeJobs(node *corev1.Node, nodeCache v1.NodeCache, jobCache ctlbatchv1.JobCache) error {
```

Comment:
sorry one question: how about the witness node? not mean to calculate it or not, but do we allow changing this on the witness node (it's not supposed to run any VMs).

Reply:
Perhaps we should avoid user to change cpu manager policy on the witness node ? I can add another validation to avoid that thing happen.